flax/experimental/nnx/nnx/nn/lora.py

```diff
@@ -0,0 +1,192 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright 2023 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import typing as tp
+
+import jax
+import jax.numpy as jnp
+
+from flax.experimental import nnx
+from flax.experimental.nnx.nnx import rnglib, variables
+from flax.experimental.nnx.nnx.module import Module
+from flax.experimental.nnx.nnx.nn import initializers
+from flax.typing import (
+  Dtype,
+  Initializer,
+)
+
+Array = jax.Array
+Axis = int
+Size = int
+A = tp.TypeVar('A')
+
+default_kernel_init = initializers.lecun_normal()
+
+
+class LoRAParam(variables.Variable[A]): pass
+
+
+class LoRA(Module):
+  """A standalone LoRA layer.
+
+  Example usage::
+
+    >>> from flax.experimental import nnx
+    >>> import jax, jax.numpy as jnp
+    >>> layer = nnx.LoRA(3, 4, 2, rngs=nnx.Rngs(0))
+    >>> layer.lora_a.value.shape
+    (3, 2)
+    >>> layer.lora_b.value.shape
+    (2, 4)
+    >>> # Wrap around existing layer
+    >>> linear = nnx.Linear(3, 4, rngs=nnx.Rngs(0))
+    >>> wrapper = nnx.LoRA(3, 4, 2, base_module=linear, rngs=nnx.Rngs(1))
+    >>> assert wrapper.base_module == linear
+    >>> wrapper.lora_a.value.shape
+    (3, 2)
+    >>> layer.lora_b.value.shape
+    (2, 4)
+    >>> y = layer(jnp.ones((16, 3)))
+    >>> y.shape
+    (16, 4)
+
+  Attributes:
+    in_features: the number of input features.
+    out_features: the number of output features.
+    lora_rank: the rank of the LoRA dimension.
+    base_module: a base module to call and substitute, if possible.
+    dtype: the dtype of the computation (default: infer from input and params).
+    param_dtype: the dtype passed to parameter initializers (default: float32).
+    precision: numerical precision of the computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrices.
+    use_lora_param_type: if yes, LoRA params will be of different Param type.
+      This makes it easier to split model state for trainable/nontrainables.
+  """
+  def __init__(
+    self,
+    in_features: int,
+    out_features: int,
+    lora_rank: int,
```

Comment:
Maybe this ordering is a bit more intuitive?
```suggestion
    in_features: int,
    lora_rank: int,
    out_features: int,
```

Reply:
Sure! LoRA is a generic layer and this is indeed more intuitive.

Note that the [PyTorch LoRALinear](https://pytorch.org/torchtune/stable/_modules/torchtune/modules/peft/lora.html) implementation is ordered `(in, rank, out)`. To avoid this being confusing, I will move `lora_rank` to a kwarg argument.